File: 2020-03/3-25/322零钱兑换/322.py

from typing import List


class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        """标准简单dp"""
        dp = [float('inf')] * (amount + 1)
        dp[0] = 0  # 初始化，0元钱没办法凑

        # 遍历每个amount
        for i in range(1, len(dp)):
            for coin in coins:
                if i >= coin:
                    dp[i] = min(dp[i], dp[i - coin] + 1)

        return dp[-1] if dp[-1] != float('inf') else -1

File: 2020-05/5-07/322零钱兑换/322.cpp


class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        // 状态转移方程：dp[i] = min(dp[i], dp[i - coin] + 1)
        vector<int> dp(amount + 1, INT_MAX);
        dp[0] = 0;

        for(int i = 1; i < dp.size(); ++i)
            for(const int& coin: coins)
                if(i - coin >= 0 && dp[i - coin] != INT_MAX)
                    dp[i] = min(dp[i], dp[i - coin] + 1);

        return dp[amount] == INT_MAX ? -1 : dp[amount];
    }
};